Handle missing $HOME and report better info to user

cburgdorf opened the pull request:
### What was wrong?

This fixes #784 where Trinity would not boot correctly if the `$HOME` variable isn't set.

### How was it fixed?

Work through all fallbacks (e.g. `XDG_TRINITY_ROOT`, `TRINITY_DATA_DIR`) until we hit `$HOME` and if `$HOME` is also missing give a clear explanation to the user.

If the user decides to provide the `TRINITY_DATA_DIR` or passes `--data-dir` but that is outside of the `XDG_TRINITY_ROOT` all paths must be pre-initialized as trinity won't attempt to create any directories outside of the `XDG_TRINITY_ROOT`.

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](https://c1.staticflickr.com/4/3435/3189870179_2dbe02c132_b.jpg)

pipermerriam commented on trinity/chains/__init__.py:
```diff
@@ -82,30 +85,32 @@ def is_database_initialized(chaindb: AsyncChainDB) -> bool:
 
     # Logfile
-    if is_under_xdg_trinity_root(chain_config.logfile_path):
+    if (not os.path.exists(chain_config.logfile_path) and
```
the `Path.exists()` method works here too, we can drop some of the `os.path` calls now that we're using `Path` objects most places.

pipermerriam commented on trinity/main.py:
```diff
@@ -70,15 +87,29 @@ def main() -> None:
+    try:
+        chain_config = ChainConfig.from_parser_args(args)
+    except AmbigiousFileSystem:
+            exit_because_ambigious_filesystem(logger)
```
Indentation looks off here.

cburgdorf commented:
@pipermerriam mind giving that another look.

The tests are all condensed into one parametrized one and I added one more test case for each method. I also added another exception class to give a more specialized message to the user when a data directory is set that is outside of the root which is ok but not if the user did not pre-init the paths themselves.
